baseplate/integration/thrift/__init__.py

```diff
@@ -42,82 +40,103 @@ class RequestContext(object):
     pass
 
 
-# TODO: exceptions in the event handler cause the connection to be abruptly
-# closed with no diagnostics sent to the client. that should be more obvious.
-class BaseplateProcessorEventHandler(TProcessorEventHandler):
-    """Processor event handler for Baseplate.
+class _ContextAwareHandler(object):
+    def __init__(self, handler, context, logger):
+        self.handler = handler
+        self.context = context
+        self.logger = logger
 
-    :param logging.Logger logger: The logger to use for error and debug logging.
+    def __getattr__(self, fn_name):
+        def call_with_context(*args, **kwargs):
+            self.logger.debug("Handling: %r", fn_name)
+
+            handler_fn = getattr(self.handler, fn_name)
+
+            span = self.context.trace
+            span.start()
+            try:
+                result = handler_fn(self.context, *args, **kwargs)
+            except (TApplicationException, TProtocolException, TTransportException):
+                # these are subclasses of TException but aren't ones that
+                # should be expected in the protocol
+                span.finish(exc_info=sys.exc_info())
+                raise
+            except TException:
+                # this is an expected exception, as defined in the IDL
+                span.finish()
+                raise
+            except:
+                # the handler crashed!
+                span.finish(exc_info=sys.exc_info())
+                raise
+            else:
+                # a normal result
+                span.finish()
+                return result
+        return call_with_context
+
+
+def _extract_trace_info(headers):
+    extracted_values = TraceInfo.extract_upstream_header_values(TRACE_HEADER_NAMES, headers)
+    flags = extracted_values.get("flags", None)
+    return TraceInfo.from_upstream(
+        int(extracted_values["trace_id"]),
+        int(extracted_values["parent_span_id"]),
+        int(extracted_values["span_id"]),
+        True if extracted_values["sampled"].decode("utf-8") == "1" else False,
+        int(flags) if flags is not None else None,
+    )
+
+
+def baseplateify_processor(processor, logger, baseplate, edge_context_factory=None):
```

Comment:
suggestions for a better name gladly accepted

Reply:
`wrap_processor_for_baseplate`?  `wrap_processor_with_baseplate`? `baseplate_wrap_processor`?